docs/compaction-merge-loops.md

```diff
@@ -0,0 +1,166 @@
+(i.e., the loop that processes the merged input stream).
+
+## A Model
+
+Each value has a type and sequence number. There are three
+value types: regular (V), tombstone (T), and prefix tombstone (PT).
+Since compaction logic cares about value types and sequence numbers,
+but not actual values, the simplified notation `<type><n>` can be used
+to represent a value with type `<type>` and sequence number `<n>`.
+
+Terminology:
+- A `value-type` is one of `V`, `T` or `PT`.
+- A `seqno` is an integer greater than 0.
+- A `value` is a 2-tuple of `(value-type, seqno)`.
+- A `key` is a string with length > 0.
+- An `entry` is a `key` with one or more `values`.
+
+Rules:
+- A kvset contains 1 or more entries.
+- A kvset cannot have two `V` or `T` entries with the same key.
+- A kvset cannot have two `PT` entries with the same key.
+
+Observations:
+- A kvset can have a `V` entry and a `PT` entry with the same key.
+
+## Condensed Input Stream
+
+We use the notion of a "condensed input stream" to be able to describe merge
+loops at a higher level.
+
+To understand what this means and why it is useful, consider the
+following two kvsets:
+```
+    Kvset 1:
+        Key    Values
+        ----   -----
+        foo1   V20, V16
+        foo2   V14
+        foo    PT18
+
+    Kvset 2:
+        Key    Values
+        ----   -----
+        foo1   T13, V12
+        foo2   V11
+
+Figure 2: Example input kvsets
+```
+
+In figure 1, step 2 produces a stream entries where each entry is a
+single entry from a single kvset.  We call this the raw intput stream.
+It would appear as follows for these example kvsets:
+```
+    get_next_raw() -->   [ foo,  [ PT18 ]]
+    get_next_raw() -->   [ foo1, [ V20, V16 ]]
+    get_next_raw() -->   [ foo1, [ T13, V12 ]]
+    get_next_raw() -->   [ foo2, [ V14 ]]
+    get_next_raw() -->   [ foo2, [ V11 ]]
+
+Figure 3: Raw input stream
+```
+
+The condensed view the same data is:
+```
+    get_next_condensed()  --> [ foo,  [ PT18 ]]
+    get_next_condensed()  --> [ foo1, [ V20, V16, T13, V12, PT18 ]]
+    get_next_condensed()  --> [ foo2, [ V14, V11, PT18 ]]
+
+Figure 4: Condensed input stream
+```
+
+Observations on the condensed view:
+
+- There has been no elimination of values.  It is simply a different
+  view of the entries contained in the two input kvsets.
+- Prefix tombstone ("foo", PT18) is associated with all keys that
+  match "foo".
+
+In the condensed view each key appears exactly once and with all of
+its values and all prefix tombstones that match the key.  The logic of
+translating from the raw stream to the condensed stream is not
+described here.
+
+## Compaction Merge Logic
+
+Using the condensed view abstraction, the general structure of the
+merge loop is simple:
+```
+1: void compact(iterator *iter, u64 horizon, bool drop_tombs) {
+2:
+3:     while (entry = iter->get_next_condensed()) {
+4:         key = entry.key();
+5:         values = entry.values();
+6:         decide which values should be kept;
```

Comment:
Not sure if it's worth pointing here that the list of values emitted in entry must be in decreasing seqno order... Or maybe that will be covered in the raw to condensed stream translation logic elsewhere...

Reply:
Added a note to address this in the future.